File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void findCombinationSum(int index, vector<int> &candidates, int target, vector<vector<int>> &result, vector<int> &ds){
        if(index == candidates.size()){
            if(target == 0){
                result.push_back(ds);
            }
            return;
        }
        if(candidates[index] <= target){
            ds.push_back(candidates[index]);
            findCombinationSum(index, candidates, target-candidates[index], result, ds);
            ds.pop_back();
        }
        findCombinationSum(index+1, candidates, target, result, ds);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> ds;
        findCombinationSum(0, candidates, target, result, ds);
        return result;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void findCombinationSum(int index, int candidates[], int target, List<List<Integer>> result, List<Integer> ds){
        if(index == candidates.length){
            if(target == 0){
                result.add(new ArrayList<>(ds));
            }
            return;
        }
        if(candidates[index]<=target){
            ds.add(candidates[index]);
            findCombinationSum(index, candidates, target-candidates[index], result, ds);
            ds.remove(ds.size()-1);
        }
        findCombinationSum(index+1, candidates, target, result, ds);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        findCombinationSum(0, candidates, target, result, new ArrayList<>());
        return result;
    }
}